### ISSUE TITLE
Batched RNN on GPU: CuArray only supports bits types

### ISSUE THREAD

**Issue by shrimpceviche** [FluxML/Zygote.jl#814]
Hi all, I encountered the 
```
ERROR: CuArray only supports bits types
```
while doing gradient calculation. 

The calculation is a simple RNN that would pick the output sequence element based on sequence length in a batched training setting.

I run the code on CPU and it worked out smoothly. It only has issues when the code and data are converted into GPU. 

Here is the code to reproduce the error using an Nvidia GTX 1050 GPU. 

```
#================ GPU select sequence length ===========================#
using Flux
using Flux.Optimise: update!, Momentum
using Plots
using Zygote

function select_last_entry_based_on_length(rnn_output, sequence_length_array)
    # based on the sequence length, select the relevant entry. 
    output_array = Zygote.Buffer([0.0], 1, size(sequence_length_array)[1])
    for i in 1:size(sequence_length_array)[1]
        _len = sequence_length_array[i]
        output_array[i] = rnn_output[_len][i]
    end
    return copy(output_array)
end

rnn_model = Chain(GRU(3, 8), Dense(8, 1, relu, initW=Flux.glorot_normal)) |> gpu
rnn_params = params(rnn_model)

function eval_model(x)
    out = rnn_model.(x)
    Flux.reset!(rnn_model[1])
    return out
end

function model_output(input, sequence_length_array)
    output = eval_model(input)
    selected_output = select_last_entry_based_on_length(output, sequence_length_array) |> gpu
    return selected_output
end

# The input sequence would have different length. 
batch_size = 12
sequence_length = 20
input = [rand(3,batch_size) for i in 1:sequence_length] |> gpu
target = ones(1,12) |> gpu
sequence_length_array = [rand(1:sequence_length) for i in 1:batch_size] |> gpu

loss(x, sequence_length_array,y) = Flux.Losses.mae(model_output(x, sequence_length_array), y)

initial_loss = loss(input, sequence_length_array, target)
println("initial_loss: $initial_loss")

opt = Momentum(0.005, 0.8)

loss_trajectory = []

for i in 1:300

grads = gradient(() -> loss(input, sequence_length_array, target), rnn_params)

for (i,p) in enumerate(rnn_params)
    if !isnothing(grads[p])
        Flux.Optimise.update!(opt, p, grads[p]) # the nothing gradient is fixed by reset! 
    end
end

loss_after = loss(input, sequence_length_array, target)
println("Iteration: $i loss_after: $loss_after")
push!(loss_trajectory, [i, loss_after])
end

plot([i[1] for i in loss_trajectory], [i[2] for i in loss_trajectory])
```

CUDA Version Info
```CUDA toolkit 11.0.3, artifact installation
CUDA driver 11.0.0

Libraries: 
- CUBLAS: 11.2.0
- CURAND: 10.2.1
- CUFFT: 10.2.1
- CUSOLVER: 10.6.0
- CUSPARSE: 11.1.1
- CUPTI: 13.0.0
- NVML: missing
- CUDNN: 8.0.2 (for CUDA 11.0.0)   
- CUTENSOR: 1.2.0 (for CUDA 11.0.0)

Toolchain:
- Julia: 1.5.0
- LLVM: 9.0.1
- PTX ISA support: 3.2, 4.0, 4.1, 4.2, 4.3, 5.0, 6.0, 6.1, 6.3, 6.4
- Device support: sm_35, sm_37, sm_50, sm_52, sm_53, sm_60, sm_61, sm_62, sm_70, sm_72, sm_75

1 device:
  0: GeForce GTX 1050 (sm_61, 3.146 GiB / 4.000 GiB available)
```

**Comment by shrimpceviche** [comment]
Please see https://discourse.julialang.org/t/zygote-cuarray-only-supports-bits-types/48910 for the solution around this issue.
